File: LeetCode/minimum-consecutive-cards-to-pick-up.java
import java.util.*;
/**
자바로 푼 버전
파이썬의 max(map(len,dic.values()))를 직접 구현해야하는 점.
ArrayList일때는 인덱스로 접근 못하고 get함수로 접근해야하는 점.

for each 구문을 써야할때를 구분해야하는점 유의하기.
 */
class Solution {
    public int minimumCardPickup(int[] cards) {
        HashMap<Integer, ArrayList<Integer>> map = new HashMap<>();

        for (int i = 0; i<cards.length; i++){
            if (map.containsKey(cards[i])){
                ArrayList<Integer> temp = map.get(cards[i]);
                temp.add(i);
                map.put(cards[i],temp);
            }
            else{
                ArrayList<Integer> temp = new ArrayList<>();
                temp.add(i);
                map.put(cards[i],temp);
            }
        }
        int maxLength = 0;
        
        int answer = Integer.MAX_VALUE;
        for (ArrayList<Integer> value: map.values()){
            maxLength = Math.max(maxLength,value.size());
        }
        if (maxLength<2) return -1;

        for (ArrayList<Integer> value: map.values()){
            for (int i=0;i<value.size()-1;i++){
                int compare = Math.abs(value.get(i)-value.get(i+1));
                answer = Math.min(answer,compare+1);
            }
        }
        return answer;
    }
}

File: LeetCode/minimum-consecutive-cards-to-pick-up.py
from collections import defaultdict
"""
cards 배열이 주어졌을때, 같은 숫자가 2번 등장할때까지의 간격이 가장 좁은 경우를 return한다.

일단 dict으로 각 카드 숫자에 대해 위치값을 list형태로 집어넣는다.
만약, list의 길이가 모두 2 미만이라면 -> 쌍이 없으므로 -1를 return.
아니라면, dic.values를 탐색하며 두 간격사이가 최소인걸 memozation하여 return 한다.
"""
class Solution:
    def minimumCardPickup(self, cards: List[int]) -> int:

        dic = defaultdict(list)

        for i in range(len(cards)):
            dic[cards[i]].append(i)
        
        if max(map(len,dic.values()))<2:
            return -1

        answer = inf

        for value in dic.values():
            for i in range(len(value)-1):
                answer = min(answer,abs(value[i]-value[i+1])+1)
        return answer
